Sync/BatchingClient.swift

```diff
@@ -0,0 +1,230 @@
+        guard (bytes + payload.bytes <= config.maxTotalBytes) &&
+              (count + payload.count < config.maxTotalRecords) else
+        {
+            return false
+        }
+
+        count += payload.count
+        bytes += payload.bytes
+        return true
+    }
+}
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private(set) var ifUnmodifiedSince: Timestamp?
+
+    private let config: InfoConfiguration
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+
+    private var batchMeta: BatchMeta
+    private var batchToken: BatchToken?
+    private var currentPayload: Payload
+    private var onCollectionUploaded: (POSTResult -> DeferredTimestamp)
+
+    // TODO: use I-U-S.
+    // Each time we do the storage operation, we might receive a backoff notification.
+    // For a success response, this will be on the subsequent request, which means we don't
+    // have to worry about handling successes and failures mixed with backoffs here.
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?,
+         uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult -> DeferredTimestamp))
+    {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.batchMeta = BatchMeta(config: config)
+        self.currentPayload = Payload(config: config)
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        return walk(records, f: addRecord)
+    }
+
+    public func addRecord(record: Record<T>) -> Success {
+        guard let line = self.serializeRecord(record) else {
+            return deferMaybe(SerializeRecordFailure(record: record))
+        }
+
+        let lineSize = line.utf8.count
+        guard lineSize < Sync15StorageClient.maxRecordSizeBytes else {
+            return deferMaybe(RecordTooLargeError(size: lineSize, guid: record.id))
+        }
+
+        let uploadRecord: UploadRecord = (record.id, line, lineSize)
+
+        // If we can add a record to the payload, all is good - we don't need to push to the server yet.
+        // Otherwise, try to push the payload (since it's full) to the server in a batch
+        if self.currentPayload.add(uploadRecord) {
+            return succeed()
+        }
+
+        // If we have a batch token, go ahead and push up the payload using the token.
+        // Otherwise, attempt to start a new batch. If we fail starting - no problem! The records would
+        // have already been sent in a single POST call.
+        guard let token = batchToken else {
+            return self.start(self.currentPayload)
```

Comment:
`start` should 'steal' `currentPayload` and (pseudo-atomically) replace `self.currentPayload` with a new empty payload. You don't want two `addRecord` calls to end up with `self.start` being called twice with the same payload…